extended/src/main/java/io/kubernetes/client/extended/controller/ControllerBuilder.java

```diff
@@ -0,0 +1,163 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+public class ControllerBuilder {
+
+  public static final int DEFAULT_WORKER_COUNT = 16;
+
+  private int workerCount = DEFAULT_WORKER_COUNT;
+
+  private String controllerName = "default-controller";
+  private RateLimitingQueue<Reconciler.Request> workQueue =
+      new DefaultRateLimitingQueue<Reconciler.Request>(Executors.newSingleThreadExecutor());
+
+  private SharedInformerFactory informerFactory;
+  private LeaderElector leaderElector;
+  private List<Supplier<Boolean>> readyFuncs = new ArrayList<>();
+  private Reconciler reconciler;
+
+  private ControllerBuilder() {
+    // Make constructor private explicitly
+  }
+
+  public static ControllerBuilder withInformerFactory(SharedInformerFactory informerFactory) {
+    ControllerBuilder builder = new ControllerBuilder();
+    builder.informerFactory = informerFactory;
+    return builder;
+  }
+
+  public <ApiType, ApiListType>
+      ControllerBuilder.ControllerWatchBuilder<ApiType, ApiListType> watch(
+          Class<ApiType> apiTypeClass) {
+    return new ControllerBuilder.ControllerWatchBuilder<>(apiTypeClass);
+  }
+
+  public ControllerBuilder withName(String controllerName) {
+    this.controllerName = controllerName;
+    return this;
+  }
+
+  public ControllerBuilder withWorkQueue(RateLimitingQueue<Reconciler.Request> workQueue) {
+    this.workQueue = workQueue;
+    return this;
+  }
+
+  public ControllerBuilder withReadyFunc(Supplier<Boolean> readyFunc) {
+    this.readyFuncs.add(readyFunc);
+    return this;
+  }
+
+  private ControllerBuilder withLeaderElector(LeaderElector leaderElector) {
+    this.leaderElector = leaderElector;
+    return this;
+  }
+
+  public ControllerBuilder withWorkerCount(int workerCount) {
+    this.workerCount = workerCount;
+    return this;
+  }
+
+  public ControllerBuilder withReconciler(Reconciler reconciler) {
+    this.reconciler = reconciler;
+    return this;
+  }
+
+  public Controller build() throws IllegalStateException {
+
+    DefaultController controller =
+        new DefaultController(
+            this.reconciler, this.workQueue, this.readyFuncs.stream().toArray(Supplier[]::new));
+
+    controller.setName(this.controllerName);
+    controller.setWorkerCount(this.workerCount);
+    controller.setWorkerThreadPool(Executors.newFixedThreadPool(this.workerCount));
+    if (reconciler == null) {
+      throw new IllegalStateException("Missing reconciler when building controller.");
+    }
+    controller.setReconciler(reconciler);
+
+    if (leaderElector != null) {
+      return new LeaderElectingController(this.leaderElector, controller);
+    }
+    return controller;
+  }
+
+  public class ControllerWatchBuilder<ApiType, ApiListType> {
```

Comment:
Let's extract this class out into it's own file.

Reply:
the point of making `ControllerWatchBuilder` a non-static nested-class of `ControllerBuilder` is to force the usage so that we can only build a controller-watch when building a controller, i.e. enclosing the usage of watch-controller into controller-builder. additionally, the watch-builder is accessing a few fields from parent like informer-factory if we move it a separate class, then there will be bunch of copies when instantiating watch-builder from controller-builder. is that convincing?